auth/src/main/java/io/grpc/auth/GoogleAuthLibraryCallCredentials.java

```diff
@@ -168,4 +188,50 @@ private static Metadata toHeaders(@Nullable Map<String, List<String>> metadata)
     }
     return headers;
   }
+
+  private static volatile boolean loadedReflection;
+  private static Class<? extends Credentials> serviceAccountClass;
+  private static Constructor<? extends Credentials> jwtConstructor;
+  private static Method getScopes;
+  private static Method getClientId;
+  private static Method getClientEmail;
+  private static Method getPrivateKey;
+  private static Method getPrivateKeyId;
+
+  private static Credentials serviceAccountToJwt(Credentials creds) {
+    try {
+      if (!loadedReflection) {
+        // Safe if run multiple times, since later runs produce equivalent results.
+        synchronized (GoogleAuthLibraryCallCredentials.class) {
+          serviceAccountClass
+              = Class.forName(SERVICE_ACCOUNT_CLASS_NAME).asSubclass(Credentials.class);
+          getScopes = serviceAccountClass.getMethod("getScopes");
+          getClientId = serviceAccountClass.getMethod("getClientId");
+          getClientEmail = serviceAccountClass.getMethod("getClientEmail");
+          getPrivateKey = serviceAccountClass.getMethod("getPrivateKey");
+          getPrivateKeyId = serviceAccountClass.getMethod("getPrivateKeyId");
+          Class<? extends Credentials> jwtClass
+              = Class.forName("com.google.auth.oauth2.ServiceAccountJwtAccessCredentials")
+              .asSubclass(Credentials.class);
+          jwtConstructor
+              = jwtClass.getConstructor(String.class, String.class, PrivateKey.class, String.class);
+          loadedReflection = true;
+        }
+      }
+      creds = serviceAccountClass.cast(creds);
+      Collection<?> scopes = (Collection<?>) getScopes.invoke(creds);
+      if (scopes.size() != 0) {
+        // Leave as-is, since the scopes may limit access within the service.
+        return creds;
+      }
+      return jwtConstructor.newInstance(
+          getClientId.invoke(creds),
+          getClientEmail.invoke(creds),
+          getPrivateKey.invoke(creds),
+          getPrivateKeyId.invoke(creds));
+    } catch (ReflectiveOperationException ex) {
+      log.log(Level.WARNING, "Failure converting service account credential to JWT", ex);
+      return creds;
```

Comment:
No. Using JWT is an optimization. We can continue using the other creds. The only reason to warn is so that someone could file a bug against us.